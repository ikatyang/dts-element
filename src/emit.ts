import * as ts from 'typescript';
import { IElement } from './element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) =>
  emit_native(transform(element), options);
