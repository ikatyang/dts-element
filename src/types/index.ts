export { IArrayType, IArrayTypeOptions, create_array_type } from './array-type';
export {
  IConstructorType,
  IConstructorTypeOptions,
  create_constructor_type,
} from './constructor-type';
export {
  IFunctionType,
  IFunctionTypeOptions,
  create_function_type,
} from './function-type';
export {
  IGeneralType,
  IGeneralTypeOptions,
  create_general_type,
} from './general-type';
export {
  IIntersectionType,
  IIntersectionTypeOptions,
  create_intersection_type,
} from './intersection-type';
export { IKeyofType, IKeyofTypeOptions, create_keyof_type } from './keyof-type';
export {
  ILiteralType,
  ILiteralTypeOptions,
  create_literal_type,
} from './literal-type';
export {
  IMappedType,
  IMappedTypeOptions,
  create_mapped_type,
} from './mapped-type';
export {} from './native-type';
export {
  IObjectType,
  IObjectTypeOptions,
  create_object_type,
} from './object-type';
export { ISubType, ISubTypeOptions, create_sub_type } from './sub-type';
export { ITupleType, ITupleTypeOptions, create_tuple_type } from './tuple-type';
export {
  ITypeofType,
  ITypeofTypeOptions,
  create_typeof_type,
} from './typeof-type';
export { IUnionType, IUnionTypeOptions, create_union_type } from './union-type';
