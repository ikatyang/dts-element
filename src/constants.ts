import * as ts from 'typescript';
import { create_native_type, INativeType } from './types/native-type';

export enum ElementKind {
  ArrayType,
  ClassDeclaration,
  ClassMember,
  ConditionalType,
  ConstructorType,
  EnumDeclaration,
  ExportDefault,
  ExportEqual,
  ExportMember,
  ExportNamed,
  ExportNamespace,
  FunctionDeclaration,
  FunctionType,
  GeneralType,
  GenericDeclaration,
  GlobalDeclaration,
  ImportEqual,
  ImportMember,
  ImportNamed,
  ImportNamespace,
  IndexSignature,
  InferType,
  InterfaceDeclaration,
  IntersectionType,
  JSDocComment, // tslint:disable-line:naming-convention
  KeyofType,
  LiteralType,
  MappedType,
  ModuleDeclaration,
  MultiLineComment,
  NamespaceDeclaration,
  NativeType,
  ObjectMember,
  ObjectType,
  ParameterDeclaration,
  SingleLineComment,
  SubType,
  TopLevelElement,
  TripleSlashReference,
  TupleType,
  TypeDeclaration,
  TypePredicate,
  TypeofType,
  UnionType,
  VariableDeclaration,
}

export const any_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword),
});
export const boolean_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword),
});
export const never_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword),
});
export const null_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword),
});
export const number_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword),
});
export const object_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword),
});
export const string_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
});
export const symbol_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.SymbolKeyword),
});
export const this_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.ThisKeyword),
});
export const undefined_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.UndefinedKeyword),
});
export const void_type = create_native_type({
  type: ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword),
});
