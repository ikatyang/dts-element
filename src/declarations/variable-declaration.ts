import * as ts from 'typescript';
import { IType } from '../collections';
import { any_type, ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';
import { add_declare_modifier_if_need } from '../utils';

export interface IVariableDeclarationOptions extends IElementOptions {
  name: string;
  type?: IType;
  export?: boolean;
  const?: boolean;
  let?: boolean;
}

export interface IVariableDeclaration
  extends IElement<ElementKind.VariableDeclaration>,
    IVariableDeclarationOptions {}

export const create_variable_declaration = (
  options: IVariableDeclarationOptions,
): IVariableDeclaration => ({
  ...create_element(ElementKind.VariableDeclaration),
  ...options,
});

export const is_variable_declaration = (
  value: any,
): value is IVariableDeclaration =>
  is_element(value) && value.kind === ElementKind.VariableDeclaration;

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_variable_declaration = (
  element: IVariableDeclaration,
  path: IElement<any>[],
) => {
  const type = transform(element.type || any_type, path) as ts.TypeNode;
  return ts.createVariableStatement(
    /* modifiers       */ add_declare_modifier_if_need(
      element.export === true
        ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
        : undefined,
      path,
    ),
    /* declarationList */ ts.createVariableDeclarationList(
      /* declarations  */ [ts.createVariableDeclaration(element.name, type)],
      /* flags         */ element.const === true
        ? ts.NodeFlags.Const
        : element.let === true ? ts.NodeFlags.Let : undefined,
    ),
  );
};
