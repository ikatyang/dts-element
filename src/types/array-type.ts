import * as ts from 'typescript';
import { IType } from '../collections';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';

export interface IArrayTypeOptions extends IElementOptions {
  type: IType;
}

export interface IArrayType
  extends IElement<ElementKind.ArrayType>,
    IArrayTypeOptions {}

export const create_array_type = (options: IArrayTypeOptions): IArrayType => ({
  ...create_element(ElementKind.ArrayType),
  ...options,
});

export const is_array_type = (value: any): value is IArrayTypeOptions =>
  is_element(value) && value.kind === ElementKind.ArrayType;

/**
 * @hidden
 */
export const transform_array_type = (
  element: IArrayType,
  path: IElement<any>[],
) =>
  ts.createArrayTypeNode(
    /* elementType */ transform(element.type, path) as ts.TypeNode,
  );
