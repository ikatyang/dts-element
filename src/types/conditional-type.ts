import * as ts from 'typescript';
import { IType } from '../collections';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';

export interface IConditionalTypeOptions extends IElementOptions {
  check: IType;
  extends: IType;
  true: IType;
  false: IType;
}

export interface IConditionalType
  extends IElement<ElementKind.ConditionalType>,
    IConditionalTypeOptions {}

export const create_conditional_type = (
  options: IConditionalTypeOptions,
): IConditionalType => ({
  ...create_element(ElementKind.ConditionalType),
  ...options,
});

export const is_conditional_type = (value: any): value is IConditionalType =>
  is_element(value) && value.kind === ElementKind.ConditionalType;

/**
 * @hidden
 */
export const transform_conditional_type = (
  element: IConditionalType,
  path: IElement<any>[],
) =>
  ts.createConditionalTypeNode(
    /* checkType   */ transform(element.check, path) as ts.TypeNode,
    /* extendsType */ transform(element.extends, path) as ts.TypeNode,
    /* trueType    */ transform(element.true, path) as ts.TypeNode,
    /* falseType   */ transform(element.false, path) as ts.TypeNode,
  );
