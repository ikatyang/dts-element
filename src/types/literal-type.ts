import * as ts from 'typescript';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';

export interface ILiteralTypeOptions extends IElementOptions {
  value: boolean | number | string;
}

export interface ILiteralType
  extends IElement<ElementKind.LiteralType>,
    ILiteralTypeOptions {}

export const create_literal_type = (
  options: ILiteralTypeOptions,
): ILiteralType => ({
  ...create_element(ElementKind.LiteralType),
  ...options,
});

export const is_literal_type = (value: any): value is ILiteralType =>
  is_element(value) && value.kind === ElementKind.LiteralType;

/**
 * @hidden
 */
export const transform_literal_type = (
  element: ILiteralType,
  _path: IElement<any>[],
) =>
  ts.createLiteralTypeNode(
    /* literal */ {
      ...ts.createLiteral(element.value),
      _literalExpressionBrand: true,
      text: element.value.toString(),
    },
  );
