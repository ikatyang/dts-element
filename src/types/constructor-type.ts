import * as ts from 'typescript';
import { IType } from '../collections';
import { any_type, ElementKind } from '../constants';
import { IGenericDeclaration } from '../declarations/generic-declaration';
import { IParameterDeclaration } from '../declarations/parameter-declaration';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';
import { create_type_parameters } from '../utils';

export interface IConstructorTypeOptions extends IElementOptions {
  return?: IType;
  generics?: IGenericDeclaration[];
  parameters?: IParameterDeclaration[];
}

export interface IConstructorType
  extends IElement<ElementKind.ConstructorType>,
    IConstructorTypeOptions {}

export const create_constructor_type = (
  options: IConstructorTypeOptions = {},
): IConstructorType => ({
  ...create_element(ElementKind.ConstructorType),
  ...options,
});

export const is_constructor_type = (value: any): value is IConstructorType =>
  is_element(value) && value.kind === ElementKind.ConstructorType;

/**
 * @hidden
 */
// tslint:disable:ter-indent
export const transform_constructor_type = (
  element: IConstructorType,
  path: IElement<any>[],
) =>
  ts.createConstructorTypeNode(
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* parameters      */ (element.parameters || []).map(
      parameter => transform(parameter, path) as ts.ParameterDeclaration,
    ),
    /* type            */ transform(
      element.return || any_type,
      path,
    ) as ts.TypeNode,
  );
