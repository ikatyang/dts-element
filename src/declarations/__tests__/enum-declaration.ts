import { any_type } from '../../constants';
import { emit } from '../../emit';
import { create_literal_type } from '../../types/literal-type';
import {
  create_enum_declaration,
  is_enum_declaration,
} from '../enum-declaration';
import { create_variable_declaration } from '../variable-declaration';

it('should return correctly with name', () => {
  expect(
    emit(
      create_enum_declaration({
        name: 'E',
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with name, const', () => {
  expect(
    emit(
      create_enum_declaration({
        name: 'E',
        const: true,
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with name, export', () => {
  expect(
    emit(
      create_enum_declaration({
        name: 'E',
        export: true,
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with name, members', () => {
  expect(
    emit(
      create_enum_declaration({
        name: 'E',
        members: [
          create_variable_declaration({
            name: 'a',
            type: create_literal_type({
              value: 5,
            }),
          }),
          create_variable_declaration({
            name: 'b',
          }),
          create_variable_declaration({
            name: 'c',
            type: create_literal_type({
              value: 'hello',
            }),
          }),
        ],
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with name, export, members', () => {
  expect(
    emit(
      create_enum_declaration({
        name: 'E',
        export: true,
        members: [
          create_variable_declaration({
            name: 'a',
            type: create_literal_type({
              value: 5,
            }),
          }),
          create_variable_declaration({
            name: 'b',
          }),
        ],
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with numerical member name', () => {
  expect(
    emit(
      create_enum_declaration({
        name: 'E',
        members: [
          create_variable_declaration({
            name: 0,
          }),
        ],
      }),
    ),
  ).toMatchSnapshot();
});

it('should throw error with members (VariableDeclaration with type (not undefined or LiteralType))', () => {
  expect(() =>
    emit(
      create_enum_declaration({
        name: 'E',
        members: [
          create_variable_declaration({
            name: 'a',
            type: any_type,
          }),
        ],
      }),
    ),
  ).toThrowError();
});

describe('is_enum_declaration', () => {
  it('should return correctly', () => {
    const element = create_enum_declaration({
      name: 'E',
    });
    expect(is_enum_declaration(element)).toBe(true);
  });
});
