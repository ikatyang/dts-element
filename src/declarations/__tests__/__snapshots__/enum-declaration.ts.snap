// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with name 1`] = `
"declare enum E {
}"
`;

exports[`should return correctly with name, const 1`] = `
"declare const enum E {
}"
`;

exports[`should return correctly with name, export 1`] = `
"export declare enum E {
}"
`;

exports[`should return correctly with name, export, members 1`] = `
"export declare enum E {
    a = 5,
    b = 6
}"
`;

exports[`should return correctly with name, members 1`] = `
"declare enum E {
    a = 5,
    b = 6,
    c = \\"hello\\"
}"
`;

exports[`should return correctly with numerical member name 1`] = `
"declare enum E {
    0 = 0
}"
`;
